File: tools/tasks/release.py
import json, tempfile, os, shutil
from tools import common
from pathlib import Path

def _extract_notes_from_changelog(version: str, changelog_path: Path) -> str:
    """
    Extract the markdown block for a version from a Keep a Changelog file.

    Accepts headings like:
      ## [0.1.2] - 2025-09-26
      ## [v0.1.2] - 2025-09-26
    and returns everything up to the next '## ' heading (or EOF).
    """
    if not changelog_path.is_file():
        common.fail(f'Error: CHANGELOG not found at "{changelog_path}".')

    text = changelog_path.read_text(encoding="utf-8")

    # Heading for this version (with optional 'v' inside the brackets, optional date)
    m = common.regex.CHANGELOG_HEADER_FOR(version).search(text)
    if not m:
        common.fail(f'Error: CHANGELOG has no section for version {version}.')
    print(" - version present in CHANGELOG.md")

    start = m.end()

    # Next version heading starts a new section
    n = common.regex.CHANGELOG_NEXT_HEADER.search(text, start)
    end = n.start() if n else len(text)

    notes = text[start:end].strip()
    if not notes:
        common.fail(f'Error: CHANGELOG section for {version} is empty.')
    print(f" - version changes in CHANGELOG.md is not empty")

    return notes

def _verify_version_link_reference(version: str, changelog_path: Path):
    """
    Verify the reference link like:
      [0.1.2]: https://github.com/<owner>/<repo>/releases/tag/v0.1.2
    exists and matches the version. Allows [v0.1.2] label too.
    """
    text = changelog_path.read_text(encoding="utf-8")

    # Find the reference line (case-insensitive on the label's 'v', URL case left as-is)
    m = common.regex.REF_LINE_URL_FOR(version).search(text)
    if not m:
        common.fail(f'Error: CHANGELOG is missing a reference link for version {version} '
             f'(e.g., "[{version}]: https://.../releases/tag/v{version}").')

    url = m.group('url')
    expected_fragment = f"/releases/tag/v{version}"
    if expected_fragment not in url:
        common.fail(f'Error: Version reference URL does not point to "{expected_fragment}".\n'
             f'Found: {url}')
    print(" - version link is present in CHANGELOG.md")


def _precheck(folder, tag_name):
    if shutil.which("gh") is None:
        common.fail("Error: GitHub CLI (gh) is not installed or not on PATH. See https://cli.github.com/")
    print(f" - gh present")

    code, out, err = common.run(["gh", "auth", "status"])
    if code != 0:
        details = err or out or "Unknown authentication error."
        common.fail(f"Error: gh is not authenticated or token is invalid/expired.\nDetails: {details}")
    print(f" - gh authenticated")

    code, out, err = common.run(["gh", "api", "user"])
    if code != 0:
        details = err or out or "Unknown API error."
        common.fail(f"Error: gh token appears invalid/expired when calling API.\nDetails: {details}")
    print(f" - gh can make requests")

    if not os.path.isdir(folder):
        common.fail(f'Error: required folder "{folder}" does not exist.')
    if not os.path.isfile(folder / "system.json"):
        common.fail(f'Error: required file "{folder / "system.json"}" does not exist.')
    if not os.path.isfile(folder / f"unofficial-FVTT-ose-{tag_name}.zip"):
        common.fail(f'Error: required file "{folder / f"unofficial-FVTT-ose-{tag_name}.zip"}" does not exist.')
    print(f" - required files and folders present")

    common.run(["git", "fetch", "--prune", "origin"])
    common.run(["git", "fetch", "--prune", "--tags", "origin"])
    print(f" - updated tags")

    #code, out, err = common.run(["git", "rev-parse", "-q", "--verify", f"refs/tags/{tag_name}"])
    # if code == 0:
    #     details = err or out or "local tag already exists."
    #     common.fail(f"Error: tag for the version already exists in local.\nDetails: {details}")
    # print(f" - no local tag with intended name")

    code, out, err = common.run(["git", "ls-remote", "--exit-code", "--refs", "--tags", "origin", f"refs/tags/{tag_name}"])
    if code == 0:
        details = err or out or "Tag already exists."
        common.fail(f"Error: tag for the version already exists.\nDetails: {details}")
    elif code not in (2, 0):
        common.fail(f"Error: unable to query remote tags.\nDetails: {err or out or 'Unknown'}")
    print(f" - no remote tag with intended name")

def run_release(version: str):
    final_version = common.normalize_version(version)
    project_root = common.project_root()
    version_folder = project_root / "releases" / final_version
    changelog_path = project_root / "CHANGELOG.md"
    print(f"Releasing version v{final_version}")
    print(f"Running prechecks...")
    _verify_version_link_reference(final_version, changelog_path)
    notes = _extract_notes_from_changelog(final_version, changelog_path)
    _precheck(version_folder, f"v{final_version}")
    print(f"Passed prechecks")
    print("")

    print(f"== patch notes =======================")
    print(f"{notes}")
    print(f"======================================")
    try:
        with tempfile.NamedTemporaryFile("w", delete=False, suffix=".md", encoding="utf-8") as tf:
            tf.write(notes)
            notes_file = tf.name

        code, out, err = common.run([
            "gh", "pr", "create",
            "--base", "stable",
            "--title", f"v{final_version}",
            "--body",  f"v{final_version}",
            "--json", "number,url"
        ])
        if code != 0:
            common.fail(f"PR creation failed: {err or out}")
        pr = json.loads(out)
        print(f"PR opened: #{pr['number']} {pr['url']}")

        # code, out, err = run(["gh", "pr", "review", str(pr["number"]), "--approve"])
        # if code != 0:
        #     print(f"Warning: PR approval failed (continuing): {err or out}", file=sys.stderr)

        code, out, err = common.run([
            "gh", "pr", "merge", str(pr["number"]),
            "--merge", "--delete-branch", "--admin", "--confirm"
        ])
        if code != 0:
            common.fail(f"PR merge failed: {err or out}")
        print("PR merged successfully.")

        # run(["git", "fetch", "origin", "stable"])
        # run(["git", "tag", "-a", f"v{version}", "origin/stable", "-m", f"v{version}"])
        # run(["git", "push", "origin", f"v{version}"])

        assets = [
            str(version_folder / "system.json"),
            str(version_folder / f"unofficial-FVTT-ose-v{final_version}.zip"),
        ]
        code, out, err = common.run([
            "gh", "release", "create", f"v{final_version}",
            "--target", "stable",
            "--title",  f"v{final_version}",
            "--notes-file", notes_file,
            "--latest",
            *assets
        ])

        if code != 0:
            common.fail(f"Release creation failed: {err or out}")

        code, out, err = common.run([
            "gh", "release", "view", f"v{final_version}",
            "--json", "url,isLatest,assets",
            "--jq", "{url,isLatest,assets:[.assets[].name]}"
        ])
        if code != 0:
            common.fail(f"Release view failed: {err or out}")
        print(out)
    finally:
        try:
            os.remove(notes_file)
        except OSError:
            pass

File: tools/tasks/release.js
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import * as common from '../common.js';

console.log(`Loaded: ${import.meta.url}`);

/**
 * Extract the markdown block for a version from a Keep a Changelog file.
 * Accepts headings like:
 *   ## [0.1.2] - 2025-09-26
 *   ## [v0.1.2] - 2025-09-26
 * and returns everything up to the next '## ' heading (or EOF).
 * @param {string} version
 * @param {string} changelogPath
 * @returns {string}
 */
function extractNotesFromChangelog(version, changelogPath) {
    if (!fs.existsSync(changelogPath)) {
        common.fail(`Error: CHANGELOG not found at "${changelogPath}".`);
    }
    const text = fs.readFileSync(changelogPath, 'utf-8');

    const headerRe = common.regex.CHANGELOG_HEADER_FOR(version);
    const m = headerRe.exec(text);
    if (!m) common.fail(`Error: CHANGELOG has no section for version ${version}.`);
    console.log(' - version present in CHANGELOG.md');

    const start = m.index + m[0].length;

    const nextHeaderRe = new RegExp(common.regex.CHANGELOG_NEXT_HEADER.source, common.regex.CHANGELOG_NEXT_HEADER.flags);
    nextHeaderRe.lastIndex = start;
    const n = nextHeaderRe.exec(text);
    const end = n ? n.index : text.length;

    const notes = text.slice(start, end).trim();
    if (!notes) common.fail(`Error: CHANGELOG section for ${version} is empty.`);
    console.log(' - version changes in CHANGELOG.md is not empty');

    return notes;
}

/**
 * Verify the reference link like:
 *   [0.1.2]: https://github.com/<owner>/<repo>/releases/tag/v0.1.2
 * exists and matches the version. Allows [v0.1.2] label too.
 * @param {string} version
 * @param {string} changelogPath
 */
function verifyVersionLinkReference(version, changelogPath) {
    const text = fs.readFileSync(changelogPath, 'utf-8');

    const m = common.regex.REF_LINE_URL_FOR(version).exec(text);
    if (!m) {
        common.fail(`Error: CHANGELOG is missing a reference link for version ${version} `
            + `(e.g., "[${version}]: https://.../releases/tag/v${version}").`);
    }

    const url = m[1];
    const expectedFragment = `/releases/tag/v${version}`;
    if (!url.includes(expectedFragment)) {
        common.fail(`Error: Version reference URL does not point to "${expectedFragment}".\nFound: ${url}`);
    }
    console.log(' - version link is present in CHANGELOG.md');
}

/**
 * @param {string} folder
 * @param {string} tagName
 */
function precheck(folder, tagName) {
    const {code: ghVersionCode} = common.run(['gh', '--version']);
    if (ghVersionCode !== 0) {
        common.fail('Error: GitHub CLI (gh) is not installed or not on PATH. See https://cli.github.com/');
    }
    console.log(' - gh present');

    let r = common.run(['gh', 'auth', 'status']);
    if (r.code !== 0) common.fail(`Error: gh is not authenticated or token is invalid/expired.\nDetails: ${r.stderr || r.stdout || 'Unknown authentication error.'}`);
    console.log(' - gh authenticated');

    r = common.run(['gh', 'api', 'user']);
    if (r.code !== 0) common.fail(`Error: gh token appears invalid/expired when calling API.\nDetails: ${r.stderr || r.stdout || 'Unknown API error.'}`);
    console.log(' - gh can make requests');

    if (!fs.existsSync(folder) || !fs.statSync(folder).isDirectory()) {
        common.fail(`Error: required folder "${folder}" does not exist.`);
    }
    if (!fs.existsSync(path.join(folder, 'system.json'))) {
        common.fail(`Error: required file "${path.join(folder, 'system.json')}" does not exist.`);
    }
    const zipPath = path.join(folder, `unofficial-FVTT-ose-${tagName}.zip`);
    if (!fs.existsSync(zipPath)) {
        common.fail(`Error: required file "${zipPath}" does not exist.`);
    }
    console.log(' - required files and folders present');

    common.run(['git', 'fetch', '--prune', 'origin']);
    common.run(['git', 'fetch', '--prune', '--tags', 'origin']);
    console.log(' - updated tags');

    r = common.run(['git', 'ls-remote', '--exit-code', '--refs', '--tags', 'origin', `refs/tags/${tagName}`]);
    if (r.code === 0) {
        common.fail(`Error: tag for the version already exists.\nDetails: ${r.stderr || r.stdout || 'Tag already exists.'}`);
    } else if (r.code !== 2) {
        common.fail(`Error: unable to query remote tags.\nDetails: ${r.stderr || r.stdout || 'Unknown'}`);
    }
    console.log(' - no remote tag with intended name');
}

/**
 * @param {string} version
 */
export async function runRelease(version) {
    const finalVersion = common.normalizeVersion(version);
    const projectRoot = common.projectRoot();
    const versionFolder = path.join(projectRoot, 'releases', finalVersion);
    const changelogPath = path.join(projectRoot, 'CHANGELOG.md');
    console.log(`Releasing version v${finalVersion}`);
    console.log('Running prechecks...');
    verifyVersionLinkReference(finalVersion, changelogPath);
    const notes = extractNotesFromChangelog(finalVersion, changelogPath);
    precheck(versionFolder, `v${finalVersion}`);
    console.log('Passed prechecks');
    console.log('');

    console.log('== patch notes =======================');
    console.log(notes);
    console.log('======================================');

    let notesFile;
    try {
        notesFile = path.join(fs.mkdtempSync(path.join(os.tmpdir(), 'tools-')), 'notes.md');
        fs.writeFileSync(notesFile, notes, 'utf-8');

        let r = common.run([
            'gh', 'pr', 'create',
            '--base', 'stable',
            '--title', `v${finalVersion}`,
            '--body', `v${finalVersion}`,
            '--json', 'number,url',
        ]);
        if (r.code !== 0) common.fail(`PR creation failed: ${r.stderr || r.stdout}`);
        const pr = JSON.parse(r.stdout);
        console.log(`PR opened: #${pr.number} ${pr.url}`);

        r = common.run([
            'gh', 'pr', 'merge', String(pr.number),
            '--merge', '--delete-branch', '--admin', '--confirm',
        ]);
        if (r.code !== 0) common.fail(`PR merge failed: ${r.stderr || r.stdout}`);
        console.log('PR merged successfully.');

        const assets = [
            path.join(versionFolder, 'system.json'),
            path.join(versionFolder, `unofficial-FVTT-ose-v${finalVersion}.zip`),
        ];
        r = common.run([
            'gh', 'release', 'create', `v${finalVersion}`,
            '--target', 'stable',
            '--title', `v${finalVersion}`,
            '--notes-file', notesFile,
            '--latest',
            ...assets,
        ]);
        if (r.code !== 0) common.fail(`Release creation failed: ${r.stderr || r.stdout}`);

        r = common.run([
            'gh', 'release', 'view', `v${finalVersion}`,
            '--json', 'url,isLatest,assets',
            '--jq', '{url,isLatest,assets:[.assets[].name]}',
        ]);
        if (r.code !== 0) common.fail(`Release view failed: ${r.stderr || r.stdout}`);
        console.log(r.stdout);
    } finally {
        try {
            if (notesFile) fs.rmSync(path.dirname(notesFile), {recursive: true, force: true});
        } catch {
            // best effort cleanup
        }
    }
}
